### ISSUE TITLE
Poor pre-training model weights

### ISSUE THREAD

**Issue by CrazyFacy** [CHENXINYAN-sg/WiFi-CSI-Sensing-Benchmark#24]
Hi, first of all thank you for your excellent work on wifi perception, but I am not getting good results when I run the pre-trained model you provided. Can you please provide effective pre-training models?
Here's my code for loading the pre-trained model, and I confirm that I should be error free:
![Image](https://github.com/user-attachments/assets/e3bedf28-9e84-4beb-a815-ffde40f7314a)

This is the effect I get:

![Image](https://github.com/user-attachments/assets/f579fadd-3036-4c64-bcae-75f6909b9385)

**Comment by CrazyFacy** [comment]
Well, I seem to have found the problem, the provided dataset code is not running with the same results on a win system as it is on a linux system causing the problem.
I use gemini to summarize this error as follows:

### Summary of the Error

1.  **Core Problem:** A machine learning model trained in a Linux environment performs with 0% accuracy when tested in a Windows environment.
2.  **Specific Manifestation:** Debugging revealed a **consistent permutation** in the predicted labels versus the true labels. For example, a true label of 0 is always predicted as 2, a true label of 1 is always predicted as 4, and so on. This indicates a systematic, not random, error.

---

### Root Cause Diagnosis

The fundamental cause of the error lies in the **inconsistency of the class-to-numeric-label mapping** between the training (Linux) and inference (Windows) environments.

Let's review the diagnostic process:

1.  **Initial Suspicion:** Differences in default file or folder sorting behavior across operating systems (Linux vs. Windows).
    * **NTU-Fi-HumanID (numeric folder names):** Expected to sort by numerical value (e.g., 1, 2, ..., 10).
    * **NTU-Fi_HAR (string folder names):** Expected to sort alphabetically (e.g., box, circle, clean...).

2.  **User-provided Windows testing code:**
    * The code uses `self.folder_paths = sorted(list(category_dirs))`.
    * Running this code on Windows showed the output: `{'box': 0, 'circle': 1, 'clean': 2, 'fall': 3, 'run': 4, 'walk': 5}`. This confirmed that the current `CSI_Dataset` code on Windows correctly sorts categories in **standard alphabetical order**.

3.  **User-provided Linux training code (Crucial Clue):**
    * The code uses `self.folder = glob.glob(root_dir + '/*/')` to get the list of folders.
    * Then, it directly constructs the `self.category` mapping using `{self.folder[i].split('/')[-2]: i ...}`.

4.  **Diagnosis of the Core Conflict: `glob.glob()`'s Non-Determinism**
    * The `glob.glob()` function **does not guarantee the order of its returned results**. Its behavior depends on the underlying operating system and file system implementation.
    * **Crucially, even though `glob.glob()` on Linux might *often* return results in lexicographical (alphabetical) order, this behavior is not strictly guaranteed.**
    * **The most critical issue is that when the model was trained on Linux, its learned label mapping was based on a specific, but unreliable, folder order returned by `glob.glob()` in that particular Linux environment.** This order was different from standard alphabetical (or numerical) order, and also different from the standard order obtained when running the same code on Windows.

5.  **Summary of the Precise Error Chain:**
    * **During Linux Training:** `glob.glob()` returned a **non-standard order** of folders (e.g., `['clean', 'run', 'box', ...]`). The model then learned the category-to-label mapping based on this non-standard order (e.g., `clean -> 0, run -> 1, box -> 2`).
    * **During Windows Testing:** The user's current `CSI_Dataset` code (or a version I suggested with `sorted()`) loads the data in **standard alphabetical order** (e.g., `['box', 'circle', 'clean', ...]`) and maps it to labels (e.g., `box -> 0, circle -> 1, clean -> 2`).
    * **Result:** When data from the `box` category is loaded in the Windows environment, it's assigned a true label of `0`. However, since the model was trained on Linux, it learned that when the data is `box`, the corresponding label should be `2`. Therefore, the model correctly predicts `2` according to its training, but when evaluated against the Windows-assigned true label `0`, it appears as a "true label 0 predicted as 2" error, leading to 0% accuracy.

---

### Solution Summary

The fundamental solution to this problem is to **ensure that the class-to-numeric-label mapping order is absolutely consistent and predictable** during both training and inference.

1.  **Recommended Solution (Most Robust): Retrain the Model**
    * **Update `CSI_Dataset`:** Use a **clear and robust `CSI_Dataset` version** that leverages `os.walk` to find all category folders and then uses `sorted()` combined with a **custom sort key (`custom_sort_key`)** to guarantee a **deterministic order** of folder paths.
        * For **purely numeric folder names** (e.g., `001`), the `custom_sort_key` should convert folder names to integers for sorting (`return int(os.path.basename(path))`).
        * For **purely string folder names** (e.g., `box`), the `custom_sort_key` should convert folder names to lowercase for alphabetical sorting (`return os.path.basename(path).lower()`).
    * **On the Linux environment, retrain your model** **using this updated `CSI_Dataset`**. This is crucial, as the model will then learn a label mapping based on a standard and deterministic order.
    * **On the Windows environment, use the** **exact same updated `CSI_Dataset`** **to load data and perform inference.**
    * **Verification:** During both Linux training and Windows testing, print out the `Detected Categories (Sorted):` and `Sorted Folder Paths for Labeling:` and **strictly compare these outputs** to ensure they are identical.

2.  **Alternative Solution (If Retraining is Not Possible): Force Windows Inference to Match Linux Training Sort**
    * **Identify the Exact Linux Training Sort Order:** On the Linux environment, run the *original* `CSI_Dataset` code (the one using `glob.glob()`) and print the actual order of `self.folder` and the resulting `self.category` mapping. This non-standard order is what your model was trained on.
    * **Hardcode into Windows Code:** **Hardcode this exact (potentially non-standard) folder order discovered on Linux** into your Windows `CSI_Dataset`. This means replacing the non-deterministic sorting with a predefined list that ensures `self.folder_paths` and the `self.category` dictionary are built according to that specific order.
    * **Verification:** Ensure that the `Sorted Folder Paths for Labeling:` printed by the `CSI_Dataset` running on Windows now perfectly matches the exact order found during Linux training.

**Final Recommendation:** While the second method might get the existing model working, the first method (retraining the model with a standardized `CSI_Dataset`) is the **more robust, maintainable, and highly recommended** solution, as it fundamentally eliminates the non-deterministic nature of data loading order.

**Comment by CrazyFacy** [comment]
In this case, I recommend rewriting the provided dataset code and re-pre-training it.
`
class CSI_Dataset(Dataset):

    def __init__(self, root_dir, modal='CSIamp', transform=None, few_shot=False, k=5, single_trace=True):
        """
        Args:
            root_dir (string): Directory with all the .mat files.
            modal (CSIamp/CSIphase): CSI data modal to load from .mat file.
            transform (callable, optional): Optional transform to be applied
                on a sample.
            few_shot (bool): Placeholder for few-shot learning logic.
            k (int): Placeholder for k-shot parameter.
            single_trace (bool): Placeholder for single trace logic.
        """
        self.root_dir = root_dir
        self.modal = modal
        self.transform = transform

        self.all_data = [] 
        self.all_labels = []  

        category_dirs = set()
        for dirpath, dirnames, filenames in os.walk(root_dir):
            for f in filenames:
                if f.endswith('.mat'):
                    category_dirs.add(dirpath)
                    break  

        self.folder_paths = sorted(list(category_dirs))

        self.category = {os.path.basename(path): i for i, path in enumerate(self.folder_paths)}

        print(f"Dataset Initializing from: {self.root_dir}")
        print(f"Detected Categories: {self.category}")
      
        print("Preloading all data, this might take a while...")

        total_files_to_load = 0
        for category_path in self.folder_paths:
            total_files_to_load += len(glob.glob(os.path.join(category_path, '*.mat')))

        if total_files_to_load == 0:
            print("WARNING: No .mat files found in the specified root directory or its subdirectories.")
            print("Please check `root_dir` and your file structure.")

        with tqdm(total=total_files_to_load, desc="Loading .mat files") as pbar:
            for category_path in self.folder_paths:
                category_name = os.path.basename(category_path)
                y_label = self.category[category_name]

                
                mat_files_in_category = glob.glob(os.path.join(category_path, '*.mat'))

                for sample_dir in mat_files_in_category:
                    try:
                        x = sio.loadmat(sample_dir)[self.modal]

                        # normalize
                        x = (x - 42.3199) / 4.9802

                        # sampling: 2000 -> 500
                        
                        if x.ndim == 2 and x.shape[1] >= 2000:
                            x = x[:, ::4]  
                            
                            if x.shape[1] == 500:
                                x = x.reshape(3, 114, 500)  
                            else:
                                print(
                                    f"WARNING: After downsampling, data for {sample_dir} has shape {x.shape}, expected second dim to be 500. Reshaping might fail.")
                                
                                x = np.zeros((3, 114, 500), dtype=np.float32)  
                        else:
                            
                            print(
                                f"WARNING: Data for {sample_dir} has unexpected shape {x.shape} before sampling. Expected 2D with second dim >= 2000.")
                            x = np.zeros((3, 114, 500), dtype=np.float32)  

                     
                        if self.transform:
                            x = self.transform(x)

                        self.all_data.append(torch.FloatTensor(x))
                        self.all_labels.append(y_label)

                    except Exception as e:
                        print(f"ERROR processing file {sample_dir} during preload: {e}")
                        self.all_data.append(torch.FloatTensor(np.zeros((3, 114, 500), dtype=np.float32)))
                        self.all_labels.append(-1)  
                    finally:
                        pbar.update(1)  

        print(f"Total samples preloaded: {len(self.all_data)}")
        if not self.all_data:
            print(
                "WARNING: No data was successfully preloaded! This might indicate an issue with file paths or contents.")

    def __len__(self):
        return len(self.all_data)

    def __getitem__(self, idx):
        x = self.all_data[idx]
        y = self.all_labels[idx]
        return x, y`

**Comment by CrazyFacy** [comment]
The same problem occurs with the widar dataset as a result:
`
class Widar_Dataset(Dataset):
    def __init__(self, root_dir):
        self.root_dir = root_dir
        self.all_data = []  
        self.all_labels = [] 
        self.folder_paths = [d for d in glob.glob(os.path.join(root_dir, '*')) if os.path.isdir(d)]

        def custom_sort_key(path):
            category_name = os.path.basename(path)
            parts = category_name.split('-', 1)
            if parts[0].isdigit():
                return int(parts[0])
            return float('inf')

        self.folder_paths.sort(key=custom_sort_key)

        self.category = {os.path.basename(path): i for i, path in enumerate(self.folder_paths)}

        print(f"Dataset Initializing from: {self.root_dir}")
        print(f"Detected Categories (Custom Sorted): {self.category}")

        print("Preloading all data, this might take a while...")
        total_files_to_load = 0
        for category_name_str in self.category:
            category_path = os.path.join(root_dir, category_name_str)
            if os.path.isdir(category_path):
                total_files_to_load += len(glob.glob(os.path.join(category_path, '*.csv')))

        with tqdm(total=total_files_to_load, desc="Loading CSV files") as pbar:
            for category_name_str, label_int in self.category.items():
                category_path = os.path.join(root_dir, category_name_str)

                if not os.path.isdir(category_path):
                    # print(f"WARNING: Category directory not found: {category_path}. Skipping.")
                    continue

                files_in_category = glob.glob(os.path.join(category_path, '*.csv'))

                for sample_dir in files_in_category:
                    try:
                        y = label_int
                        x = np.genfromtxt(sample_dir, delimiter=',')

                        if x.size == 0:
                            # print(f"WARNING: CSV file is empty or could not be read: {sample_dir}")
                            x = np.zeros((22, 20, 20), dtype=np.float32)
                        elif np.all(x == 0):
                            pass

                        x = (x - 0.0025) / 0.0119

                        expected_elements = 22 * 20 * 20
                        if x.size != expected_elements:
                            # print(f"ERROR: Data size mismatch for reshape in {sample_dir}.")
                            # print(f"  Expected {expected_elements} elements (22*20*20), but got {x.size}.")
                            # print(f"  Original shape of x after genfromtxt: {x.shape}")
                            x = np.zeros((22, 20, 20), dtype=np.float32)
                        else:
                            x = x.reshape(22, 20, 20)

                        self.all_data.append(torch.FloatTensor(x))
                        self.all_labels.append(y)

                    except Exception as e:
                        # print(f"ERROR processing file {sample_dir} during preload: {e}")
                        self.all_data.append(torch.FloatTensor(np.zeros((22, 20, 20), dtype=np.float32)))
                        self.all_labels.append(-1)
                    finally:
                        pbar.update(1)

        print(f"Total samples preloaded: {len(self.all_data)}")

        if not self.all_data:
            print("WARNING: No data was successfully preloaded! Please check 'root_dir' and file structure.")
        if not self.folder_paths:
            print("WARNING: No category folders were found! Ensure 'root_dir' contains subfolders for categories.")

    def __len__(self):
        return len(self.all_data)

    def __getitem__(self, idx):
        x = self.all_data[idx]
        y = self.all_labels[idx]
        return x, y`
